src/validation/recursive_validator.rs

```diff
@@ -46,16 +48,21 @@ impl RecursiveValidator {
 
     /// Entry point of finding and reporting all recursive data structures.
     pub fn validate_recursion(&mut self, index: &Index) {
-        let mut nodes_all: IndexSet<&str> = IndexSet::new();
+        let mut nodes_all: IndexSet<&DataType> = IndexSet::new();
         let mut nodes_visited = IndexSet::new();
 
         // Structs (includes arrays defined in structs)
-        nodes_all.extend(
-            index.get_types().values().filter(|x| x.get_type_information().is_struct()).map(|x| x.get_name()),
-        );
+        nodes_all.extend(index.get_types().values().filter(|x| x.get_type_information().is_struct()));
 
         // Function Blocks
-        nodes_all.extend(index.get_pous().values().filter(|x| x.is_function_block()).map(|x| x.get_name()));
+        nodes_all.extend(
+            index.get_pou_types().values().filter(|x| x.get_type_information().is_struct()).filter(|x| {
```

Comment:
the first filter (x.get_type_information().is_struct()) looks redundant to me since you do another filter with the match-statement?